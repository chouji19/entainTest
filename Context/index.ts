

export * from './RaceContext';
export * from './RaceProvider';
export * from './raceReducer';
