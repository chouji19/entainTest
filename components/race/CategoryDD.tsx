import React, { useState } from 'react';
import { StyleSheet, Text, View } from 'react-native';
import { Dropdown } from 'react-native-element-dropdown';
import { bgColor, mainColor } from '../../Constants';

const data = [
	{ label: 'All categories', value: '' },
	{ label: 'Greyhound racing', value: '9daef0d7-bf3c-4f50-921d-8e818c60fe61' },
	{ label: 'Harness racing', value: '161d9be2-e909-4326-8c2c-35ed71fb460b' },
	{ label: 'Horse racing', value: '4a2788f8-e825-4d36-9894-efd4baf1cfae' },
];

interface Props {
	onChangeHandler: (item: string) => void;
}

const CategoryDropDown: React.FC<Props> = ({ onChangeHandler }) => {
	const [value, setValue] = useState('');
	const [isFocus, setIsFocus] = useState(false);

	const renderLabel = () => {
		return (
			<Text style={[styles.label, isFocus && { color: mainColor }]}>
				Filter by category
			</Text>
		);
	};

	// item: any based on the library definition
	const onChange = (item: any) => {
		setValue(item.value);
		setIsFocus(false);
		onChangeHandler(item.value as string);
	};


	return (
		<View style={styles.container}>
			{renderLabel()}
			<Dropdown
				style={[styles.dropdown, isFocus && { borderColor: mainColor }]}
				placeholderStyle={styles.placeholderStyle}
				selectedTextStyle={styles.selectedTextStyle}
				iconStyle={styles.iconStyle}
				data={data}
				maxHeight={300}
				labelField="label"
				valueField="value"
				placeholder={!isFocus ? 'Select item' : '...'}
				searchPlaceholder="Search..."
				value={value}
				onFocus={() => setIsFocus(true)}
				onBlur={() => setIsFocus(false)}
				onChange={onChange}
			/>
		</View>
	);
};

const styles = StyleSheet.create({
	container: {
		width: '100%',
		paddingVertical: 16,
	},
	dropdown: {
		height: 50,
		borderColor: 'gray',
		borderWidth: 0.5,
		borderRadius: 8,
		paddingHorizontal: 8,
	},
	icon: {
		marginRight: 5,
	},
	label: {
		position: 'absolute',
		backgroundColor: bgColor,
		left: 22,
		top: 8,
		zIndex: 999,
		paddingHorizontal: 8,
		fontSize: 14,
	},
	placeholderStyle: {
		fontSize: 16,
	},
	selectedTextStyle: {
		fontSize: 16,
	},
	iconStyle: {
		width: 20,
		height: 20,
	},
});

export default CategoryDropDown;
