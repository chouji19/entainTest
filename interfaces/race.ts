
//Interface for the race in the app
export interface IRace {
	race_id: string;
	meeting_name: string;
	race_number: number;
	advertised_start_seconds: number;
	category: string;
}

// Interfaces from de API
export interface RaceResponse {
	status: number;
	data: Data;
	message: string;
}

export interface Data {
	next_to_go_ids: string[];
	race_summaries: { [key: string]: RaceSummary };
}

export interface RaceSummary {
	race_id: string;
	race_name: string;
	race_number: number;
	meeting_id: string;
	meeting_name: string;
	category_id: string;
	advertised_start: AdvertisedStart;
	race_form: RaceForm;
	venue_id: string;
	venue_name: string;
	venue_state: string;
	venue_country: string;
}

export interface AdvertisedStart {
	seconds: number;
}

export interface RaceForm {
	distance: number;
	distance_type: DistanceType;
	distance_type_id: string;
	track_condition: DistanceType;
	track_condition_id: string;
	weather: DistanceType;
	weather_id: string;
	race_comment: string;
	additional_data: string;
	generated: number;
	silk_base_url: string;
	race_comment_alternative: string;
}

export interface DistanceType {
	id: string;
	name: string;
	short_name: string;
	icon_uri?: string;
}
