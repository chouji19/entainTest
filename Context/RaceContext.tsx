import { createContext } from 'react';
import { IRace } from '../interfaces/race';


interface ContextProps {
	races: IRace[];
	isLoading: boolean;
	error: string;
	removeRace: (race_id: string) => Promise<void>;
	getRaces: () => Promise<void>;
}


export const RaceContext = createContext({} as ContextProps );
