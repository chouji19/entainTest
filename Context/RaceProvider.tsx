import React,{ FC, useEffect, useReducer, useState } from 'react';
import backApi from '../api/backApi';
import { IRace, RaceResponse } from '../interfaces/race';
import { RaceContext, RaceReducer } from './';

export interface RaceState {
	races: IRace[];
	isLoading: boolean,
	error: string,
}

// initial state
export const Race_INITIAL_STATE: RaceState = {
	races: [],
	isLoading: false,
	error: '',
};

export const RaceProvider:FC<{children: React.ReactNode}> = ({ children }) => {

	const [state, dispatch] = useReducer( RaceReducer , Race_INITIAL_STATE );

	// Get races and convert to an array of IRace
	const getRaces = async () => {
		dispatch({type: '[Race] - change loading state', payload: { isLoading: true }});
		try {
			const racesResult = await backApi.get<RaceResponse>('/racing/?method=nextraces&count=10');
			const racesData: IRace[] = [];

			// convert the races response to an Array to make it easier to handle
			for (let index = 0; index < racesResult.data.data.next_to_go_ids.length; index++) {
				const raceId = racesResult.data.data.next_to_go_ids[index];
				const raceTemp = racesResult.data.data.race_summaries[raceId];
				if (raceTemp) {
					racesData.push({
						advertised_start_seconds: raceTemp.advertised_start.seconds,
						category: raceTemp.category_id,
						meeting_name: raceTemp.meeting_name,
						race_id: raceTemp.race_id,
						race_number: raceTemp.race_number,
					});
				}
			}
			// sort and get 5 races
			const races: IRace[] = racesData.sort((a, b) => {
				return a.advertised_start_seconds - b.advertised_start_seconds;
			}).slice(0,5);
			dispatch({type: '[Race] - InitRaces', payload: {races }});
		} catch (err) {
			dispatch({type: '[Race] - setError', payload: {error: err as string}});

			// Clean the error to avoid displaying multiple times
			setTimeout(() => {
				dispatch({type: '[Race] - clean error'});
			}, 3000);
		} finally {
			dispatch({type: '[Race] - change loading state', payload: { isLoading: false }});
		}
	};
	useEffect(() => {
		getRaces();
	}, []);

	const removeRace = async (race_id: string) => {
		dispatch({type:  '[Race] - RemoveRace', payload: { race_id }});
	};

	return (
		<RaceContext.Provider value={{
			...state,
			getRaces,
			removeRace,
		}}>
			{ children }
		</RaceContext.Provider>
	);
};
