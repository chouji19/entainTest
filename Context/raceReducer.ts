import { IRace } from '../interfaces/race';
import { RaceState } from './';


type RaceActionType =
	| { type: '[Race] - InitRaces', payload: { races: IRace[] } }
	| { type: '[Race] - RemoveRace', payload: { race_id: string } }
	| { type: '[Race] - change loading state', payload: { isLoading: boolean } }
	| { type: '[Race] - setError', payload: { error: string } }
	| { type: '[Race] - clean error' }


export const RaceReducer = (state: RaceState, action: RaceActionType): RaceState => {

	switch (action.type) {
		case '[Race] - InitRaces':
			return {
				...state,
				races: action.payload.races,
			};
		// to remove, just filter races excluding the current one
		case '[Race] - RemoveRace':
			return {
				...state,
				races: state.races.filter(x=> x.race_id !== action.payload.race_id),
			};
		case '[Race] - change loading state':
			return {
				...state,
				isLoading: action.payload.isLoading,
			};
		case '[Race] - clean error':
			return {
				...state,
				error: '',
			};
		case '[Race] - setError':
			return {
				...state,
				error: action.payload.error,
			};

		default:
			return state;
	}

};
