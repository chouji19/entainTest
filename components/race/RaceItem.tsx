import React, { useContext, useEffect, useState } from 'react';
import { StyleSheet, View, Text } from 'react-native';
import moment from 'moment';
import { mainColor } from '../../Constants';
import { IRace } from '../../interfaces/race';
import { RaceContext } from '../../Context';

interface Props {
	race: IRace
}

const RaceItem: React.FC<Props> = ({ race }) => {
	const [timer, setTimer] = useState(race ? race.advertised_start_seconds : 60);

	const { removeRace } = useContext(RaceContext);

	const id = React.useRef<number | null>(null);
	const clear = () => {
		if (id.current) {
			clearInterval(id.current);
		}
	};

	useEffect(() => {
		id.current = setInterval(() => {
			setTimer((time) => time - 1);
		}, 1000);
		return () => clear();
	}, []);

	useEffect(() => {
		if (timer === 0) {
			removeRace(race.race_id);
			clear();
		}
	}, [race.race_id, removeRace, timer]);


	if (!race) {
		return <></>;
	}

	return (
		<View style={styles.item}>
			<View style={styles.itemInfo}>
				<Text style={styles.meetingName}>{race.meeting_name}</Text>
				<Text style={styles.raceNumber}>Race #{race.race_number}</Text>
			</View>
			<View style={styles.itemCounter}>
				<Text style={styles.countDown}>{moment.utc(timer * 1000).format('HH:mm:ss')}</Text>
			</View>
		</View>
	);
};

const styles = StyleSheet.create({
	item: {
		flexDirection: 'row',
		justifyContent: 'space-between',
		backgroundColor: '#e1e1e1',
		height: 60,
		width: '100%',
		marginVertical: 8,
		paddingHorizontal: 8,
		borderRadius: 8,
		shadowColor: mainColor,
		shadowOffset: {
			width: 0,
			height: 2,
		},
		shadowOpacity: 0.25,
		shadowRadius: 3.84,

		elevation: 5,
	},
	itemInfo: {
		flex: 1,
		justifyContent: 'center',
	},
	itemCounter: {
		justifyContent: 'center',
		marginRight: 4,
	},
	meetingName: {
		fontSize: 15,
	},
	raceNumber: {
		fontSize: 12,
		fontWeight: '300',
	},
	countDown: {
		fontSize: 35,
	},
});

export default RaceItem;
