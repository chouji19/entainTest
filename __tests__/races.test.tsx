import React from 'react';
import { act, create } from 'react-test-renderer';
import { IRace } from '../interfaces/race';
import RaceItem from '../components/race/RaceItem';
import { RaceContext } from '../Context';
import { RaceProvider } from '../Context/RaceProvider';

const race: IRace = {
	race_id: '1d1f9dbf-672d-40dc-bf88-878c486516cc',
	advertised_start_seconds: 1668732314,
	category: '9daef0d7-bf3c-4f50-921d-8e818c60fe61',
	meeting_name: 'Tristate',
	race_number: 4,
};

beforeEach(() => {
	jest.useFakeTimers();
});

const initialState = {
	error: '',
	races: [],
	isLoading: false,
	getRaces: jest.fn(),
	removeRace: jest.fn(),
};

describe('<RaceItem />', () => {
	test('renders correctly', () => {
		const component = create(
			<RaceContext.Provider value={initialState}>
				<RaceItem race={race} />
			</RaceContext.Provider>
		).toJSON();
		expect(component).toMatchSnapshot();
	});
});
