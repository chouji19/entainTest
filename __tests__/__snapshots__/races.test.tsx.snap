// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RaceItem /> renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#e1e1e1",
      "borderRadius": 8,
      "elevation": 5,
      "flexDirection": "row",
      "height": 60,
      "justifyContent": "space-between",
      "marginVertical": 8,
      "paddingHorizontal": 8,
      "shadowColor": "#BF2BFF",
      "shadowOffset": Object {
        "height": 2,
        "width": 0,
      },
      "shadowOpacity": 0.25,
      "shadowRadius": 3.84,
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 15,
        }
      }
    >
      Tristate
    </Text>
    <Text
      style={
        Object {
          "fontSize": 12,
          "fontWeight": "300",
        }
      }
    >
      Race #
      4
    </Text>
  </View>
  <View
    style={
      Object {
        "justifyContent": "center",
        "marginRight": 4,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 35,
        }
      }
    >
      00:45:14
    </Text>
  </View>
</View>
`;
